// @nimble-way/nimble-js/resources/agent.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { path } from "../internal/utils/path.mjs";
export class Agent extends APIResource {
    /**
     * List Agent Templates
     *
     * @example
     * ```ts
     * const agents = await client.agent.list();
     * ```
     */
    list(query = {}, options) {
        return this._client.get('/v1/agents', { query, ...options });
    }
    /**
     * Create Agent Generation
     *
     * @example
     * ```ts
     * const response = await client.agent.generate({
     *   prompt: 'prompt',
     *   url: 'url',
     * });
     * ```
     */
    generate(body, options) {
        return this._client.post('/v1/agents/generations', { body, ...options });
    }
    /**
     * Get Agent Template
     *
     * @example
     * ```ts
     * const agent = await client.agent.get('template_name');
     * ```
     */
    get(templateName, options) {
        return this._client.get(path `/v1/agents/${templateName}`, options);
    }
    /**
     * Get Agent Generation
     *
     * @example
     * ```ts
     * const response = await client.agent.getGeneration(
     *   '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     * );
     * ```
     */
    getGeneration(generationID, options) {
        return this._client.get(path `/v1/agents/generations/${generationID}`, options);
    }
    /**
     * Publish Agent Version
     *
     * @deprecated
     */
    publish(agentName, body, options) {
        return this._client.post(path `/v1/agents/${agentName}/publish`, { body, ...options });
    }
    /**
     * Execute WSA Realtime Endpoint
     *
     * @example
     * ```ts
     * const response = await client.agent.run({
     *   agent: 'agent',
     *   params: { foo: 'bar' },
     * });
     * ```
     */
    run(body, options) {
        return this._client.post('/v1/agents/run', { body, ...options });
    }
    /**
     * Execute WSA Async Endpoint
     *
     * @example
     * ```ts
     * const response = await client.agent.runAsync({
     *   agent: 'agent',
     *   params: { foo: 'bar' },
     * });
     * ```
     */
    runAsync(body, options) {
        return this._client.post('/v1/agents/async', { body, ...options });
    }
    /**
     * Execute WSA Batch Endpoint
     *
     * @example
     * ```ts
     * const response = await client.agent.runBatch({
     *   inputs: [{}],
     *   shared_inputs: { agent: 'agent' },
     * });
     * ```
     */
    runBatch(body, options) {
        return this._client.post('/v1/agents/batch', { body, ...options });
    }
}
